rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -230,97 +90,77 @@ class Publisher : public PublisherBase
-      rcl_interfaces::msg::IntraProcessMessage ipm;
-      ipm.publisher_id = intra_process_publisher_id_;
-      ipm.message_sequence = message_seq;
-      auto status = rcl_publish(&intra_process_publisher_handle_, &ipm);
-      if (RCL_RET_PUBLISHER_INVALID == status) {
-        rcl_reset_error();  // next call will reset error message if not context
-        if (rcl_publisher_is_valid_except_context(&intra_process_publisher_handle_)) {
-          rcl_context_t * context = rcl_publisher_get_context(&intra_process_publisher_handle_);
-          if (nullptr != context && !rcl_context_is_valid(context)) {
-            // publisher is invalid due to context being shutdown
-            return;
-          }
-        }
-      }
-      if (RCL_RET_OK != status) {
-        rclcpp::exceptions::throw_from_rcl_error(status, "failed to publish intra process message");
-      }
-    } else {
-      // Always destroy the message, even if we don't consume it, for consistency.
       msg.reset();
+      return;
+    }
+    std::unique_ptr<MessageT, MessageDeleter> msg_copy;
+    // If interprocess subscriptors exists, then promote to shared and publish.
+    if (get_subscription_count() > get_intra_process_subscription_count()) {
+      MessageSharedPtr shared_msg =
+        MessageSharedPtr(std::move(msg));
+      return this->publish(shared_msg);
     }
+    uint64_t message_seq =
+      store_intra_process_message(intra_process_publisher_id_, msg);
+    this->do_intra_process_publish(message_seq);
   }
 
   virtual void
   publish(const std::shared_ptr<MessageT> & msg)
   {
     // Avoid allocating when not using intra process.
-    if (!store_intra_process_message_) {
+    if (!intra_process_is_enabled_) {
       // In this case we're not using intra process.
       return this->do_inter_process_publish(msg.get());
     }
     // Otherwise we have to allocate memory in a unique_ptr and pass it along.
-    // TODO(wjwwood):
-    //   The intra process manager should probably also be able to store
-    //   shared_ptr's and do the "smart" thing based on other intra process
-    //   subscriptions. For now call the other publish().
+    // NOTE(ivanpauno): As the message is not const, a copy should be made.
+    // unique_ptr<MessageT> or shared_ptr<const MessageT> siganture could
+    // be used here. I decided to use the first one.
     auto ptr = MessageAllocTraits::allocate(*message_allocator_.get(), 1);
-    MessageAllocTraits::construct(*message_allocator_.get(), ptr, *msg.get());
-    MessageUniquePtr unique_msg(ptr, message_deleter_);
-    return this->publish(unique_msg);
+    MessageAllocTraits::construct(*message_allocator_.get(), ptr, *msg);
+    MessageUniquePtr unique_msg = MessageUniquePtr(ptr, message_deleter_);
+    this->publish(unique_msg);
   }
 
   virtual void
-  publish(std::shared_ptr<const MessageT> msg)
+  publish(const std::shared_ptr<MessageT> && msg)
```

Comment:
what is the purpose of a `const` rvalue ref?

Reply:
I don't understand either. `const` reference would be ok by me (though others have a dogmatic belief that you should never take a reference to a shared pointer), and I guess `&&` would be ok, but the `const &&` is odd. I suppose, like `const std::shared_ptr<T>` is implies this function will not modify it, but I'm honestly not sure if that makes sense or not. I'd have to think about it more. @ivanpauno do you have a specific reason for it?

Reply:
We may want to drop this signature all together or change it to `const std::shared_ptr<const MessageT> &`?

See: https://github.com/ros2/rclcpp/issues/640

Reply:
`publish(const std::shared_ptr<MessageT> && msg)` the extra const is unnecessary,  and it should be deleted as I now realize that it's a bad idea.

I also left two other `publish` methods taking a `shared_ptr`:
`publish(const std::shared_ptr<const MessageT> & msg)` (1)
`publish(const std::shared_ptr<MessageT> & msg)` (2)

(1) Is the one that allows sharing the ownership of the published message with the subscribers.
(2) Constructs a unique_ptr copy and stores it.
According to #640, (2) could be deleted, and (1) could be modified to `publish(std::shared_ptr<const MessageT> msg)`

Reply:
To be clearer, you could still share ownership of the message between multiple subscriptions using the `publish` method with `unique_ptr` signature, as it's internally promoted.
Creating a `std::shared_ptr<const MessageT>` is maybe a little painful, as the constructor of the messages doesn't allow initializing its members. I don't know if someone would use that `publish` signature.

Reply:
> Creating a std::shared_ptr<const MessageT> is maybe a little painful, as the constructor of the messages doesn't allow initializing its members. I don't know if someone would use that publish signature.

You don't have to create the message const, it just tells the user that we won't mutate the message inside of publish.

Consider this:

```c++
#include <cstdio>
#include <memory>
#include <string>

struct Message {
  int a;
  double b;
};

void publish1(const std::shared_ptr<const Message> msg) {
  printf(
    "publishing with a=%s and b=%s\n",
    std::to_string(msg->a).c_str(),
    std::to_string(msg->b).c_str());
}

void publish2(const std::shared_ptr<Message> msg) {
  printf(
    "publishing with a=%s and b=%s\n",
    std::to_string(msg->a).c_str(),
    std::to_string(msg->b).c_str());
}

int main() {
  {
    auto msg = std::make_shared<Message>();
    msg->a = 42;
    msg->b = 3.14;

    publish1(msg);
    publish2(msg);
  }

  {
    auto const_msg = std::make_shared<const Message>();
    // const_msg->a = 42;
    // ^ compier error:
    //     cannot assign to return value because function 'operator->' returns a const value
    // std::shared_ptr<Message> msg = const_msg;
    //                              ^ compiler error:
    //     no viable conversion from 'shared_ptr<const Message>' to 'shared_ptr<Message>'
    publish1(const_msg);  // Ok.
    // publish2(const_msg);  // Not ok, same as most recent error above.
  }

  return 0;
}
```

I'd say we should only have `publish1` and not `publish2`.

Reply:
I think `publish1` is dangerous. Because the caller could modify the message after doing the `publish` call.

I will prefer deleting all the `shared_ptr` `publish` calls, and only left the `unique_ptr` call.
The internal upgrading logic will still allow sharing the message between many subscriptions.

When having a `shared_ptr` message, the standard way of publishing it would be:

```cpp
publish(*shared_msg)
```
Which was also suggested by #640.
Obviously, we could still let a `shared_ptr` publish call, which copies the message to a `unique_ptr` and does the `unique_ptr` publish.

Reply:
The point of the `publish1` signature is not to keep the same shared pointer, it is to copy it. Obviously you cannot keep a copy of the shared pointer the user gives you, because they could modify it afterwards. So the `publish1` signature is not dangerous so long as you make a copy of the data and not just the shared pointer.

It's only a convenience function beyond `publish(const T &)`, so you don't have to do `publish(*shared_msg)`.